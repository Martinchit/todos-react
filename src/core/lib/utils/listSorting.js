export const listSortingByCompletion = (key) => (a, b) => a[key] - b[key];
