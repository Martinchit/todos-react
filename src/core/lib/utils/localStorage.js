export const retrieveToDos = () => {
  const result = localStorage.getItem(process.env.REACT_APP_TODO_LOCAL_STORAGE_KEY);
  return result ? JSON.parse(result) : null;
};

export const updateToDos = (toDos) => {
  const stringified = JSON.stringify(toDos);
  localStorage.setItem(process.env.REACT_APP_TODO_LOCAL_STORAGE_KEY, stringified);
};
