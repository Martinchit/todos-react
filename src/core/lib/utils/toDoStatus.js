export const TO_DO_STATUS = {
  ALL: 'all',
  COMPLETED: 'completed',
  ACTIVE: 'active',
};
