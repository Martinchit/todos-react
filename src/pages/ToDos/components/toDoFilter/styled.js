import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { IoCheckmarkDoneCircleSharp, IoAlertCircleSharp } from 'react-icons/io5';
import { Box } from '../../../../core/components/atoms/Box';

export const ToDoFilterContainer = styled(Box)`
  box-shadow: ${({ theme }) => `0 -5px 5px -7px ${theme.colors.black}`};
`;

export const ToDoStatsDoneIcon = styled(IoCheckmarkDoneCircleSharp)`
  color: ${({ theme }) => theme.colors.success};
  font-size: 1.5rem;
  margin-right: 3px;
`;

export const ToDoStatsActiveIcon = styled(IoAlertCircleSharp)`
  color: ${({ theme }) => theme.colors.warning};
  font-size: 1.5rem;
  margin-right: 3px;
`;

export const ToDoFilterOption = styled(Link)`
  margin: 0 3px;
  border-bottom: ${({ $selected, theme }) =>
    $selected ? `1px solid ${theme.colors.primary}` : 'none'};
  font-weight: ${({ $selected }) => ($selected ? 800 : 400)};
  color: ${({ theme }) => theme.colors.black};
  text-decoration: none;
  padding: 0 10px;
  cursor: pointer;
`;

export const CompletedToDoDeleteButton = styled.button`
  color: ${({ theme }) => theme.colors.danger};
  cursor: pointer;
  font-weight: 600;
`;
