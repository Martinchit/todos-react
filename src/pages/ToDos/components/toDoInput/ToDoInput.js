import React from 'react';
import PropTypes from 'prop-types';
import { ToDoInputField, DeleteIcon, ExpandIcon, HideIcon } from './styled';
import { Box } from '../../../../core/components/atoms/Box';
import { ENTER_KEY } from '../../../../core/lib/utils/inputEventKey';

export const ToDoInput = ({
  hasToDos,
  isExpanding,
  toggleListView,
  onToDoInsert,
}) => {
  const [toDoInput, setToDoInput] = React.useState('');

  const onInputChange = React.useCallback(
    ({ target }) => {
      setToDoInput(target.value);
    },
    [setToDoInput]
  );

  const emptyInput = React.useCallback(() => {
    setToDoInput('');
  }, []);

  const onKeyUp = React.useCallback(
    ({ code }) => {
      if (code === ENTER_KEY) {
        onToDoInsert(toDoInput.trim());
        emptyInput();
      }
    },
    [emptyInput, onToDoInsert, toDoInput]
  );

  return (
    <Box justifyContent="space-between" $elevation>
      {isExpanding ? (
        <HideIcon $visible={hasToDos} onClick={hasToDos ? toggleListView : null} />
      ) : (
        <ExpandIcon $visible={hasToDos} onClick={hasToDos ? toggleListView : null} />
      )}
      <ToDoInputField
        onChange={onInputChange}
        value={toDoInput}
        type="text"
        placeholder="What needs to be done"
        onKeyUp={onKeyUp}
      />
      <DeleteIcon $visible={toDoInput.length > 0} onClick={emptyInput} />
    </Box>
  );
};

ToDoInput.propTypes = {
  hasToDos: PropTypes.bool,
  isExpanding: PropTypes.bool,
  toggleListView: PropTypes.func,
  onToDoInsert: PropTypes.func,
};

ToDoInput.defaultProps = {
  hasToDos: false,
  isExpanding: false,
  toggleListView: null,
  onToDoInsert: null,
};
