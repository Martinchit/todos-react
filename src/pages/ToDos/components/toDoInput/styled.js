import styled from 'styled-components';
import { TiDeleteOutline, TiArrowSortedUp, TiArrowSortedDown } from 'react-icons/ti';

export const ToDoInputField = styled.input`
  width: 80%;
  padding: 2%;
  background-color: transparent;

  :focus,
  :active,
  :hover {
    outline: none;
  }

  ::placeholder,
  ::-webkit-input-placeholder {
    font-style: italic;
  }
  :-ms-input-placeholder {
    font-style: italic;
  }
`;

export const DeleteIcon = styled(TiDeleteOutline)`
  font-size: 1.5rem;
  margin: 0 3%;
  cursor: pointer;
  visibility: ${({ $visible }) => ($visible ? 'visible' : 'hidden')};
  color: ${({ theme }) => theme.colors.danger};
`;

export const ExpandIcon = styled(TiArrowSortedDown)`
  font-size: 1.5rem;
  margin: 0 3%;
  cursor: pointer;
  visibility: ${({ $visible }) => ($visible ? 'visible' : 'hidden')};
  color: ${({ theme }) => theme.colors.primary};
`;

export const HideIcon = styled(TiArrowSortedUp)`
  font-size: 1.5rem;
  margin: 0 3%;
  cursor: pointer;
  visibility: ${({ $visible }) => ($visible ? 'visible' : 'hidden')};
  color: ${({ theme }) => theme.colors.black};
`;
