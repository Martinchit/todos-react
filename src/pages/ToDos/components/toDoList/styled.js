import styled from 'styled-components';
import { RiDeleteBinLine } from 'react-icons/ri';
import { Box } from '../../../../core/components/atoms/Box';

export const ToDoItemContainer = styled(Box)`
  box-shadow: ${({ theme }) => `0 -5px 5px -7px ${theme.colors.black}`};
`;

export const ToDoLabel = styled.label`
  width: 80%;
  cursor: pointer;
`;

export const ToDoEditInput = styled.input`
  height: 100%;
  width: 86%;
  background-color: transparent;

  :focus,
  :active,
  :hover {
    outline: none;
  }

  :focus {
    border-bottom: ${({ theme }) => `2px solid ${theme.colors.success}`};
  }
`;

export const ItemDeleteButton = styled(RiDeleteBinLine)`
  font-size: 1.5rem;
  cursor: pointer;
  color: ${({ theme }) => theme.colors.danger};
`;
