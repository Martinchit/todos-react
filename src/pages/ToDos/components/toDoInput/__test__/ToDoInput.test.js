import React from "react";
import { shallow } from 'enzyme'; 
import 'jest-styled-components';

import { ToDoInput } from "../ToDoInput";
import { ToDoInputField, DeleteIcon, HideIcon, ExpandIcon } from '../styled';
import { ENTER_KEY } from '../../../../../core/lib/utils/inputEventKey';

describe("ToDoInput", () => {
  const onToDoInsert = jest.fn();
  const value = 'test';
  const isExpanding = true;
  const toggleListView = jest.fn();

  it("renders correctly", () => {
    const wrapper = shallow(<ToDoInput />)
    expect(wrapper).toMatchSnapshot()
  })

  it('handles input onChange', () => {
    const wrapper = shallow(<ToDoInput onToDoInsert={onToDoInsert} />)
    expect(wrapper.find(DeleteIcon).at(0).props()['$visible']).toBeFalsy()
    wrapper.find(ToDoInputField).at(0).simulate('change', { target: { value } })
    expect(wrapper.find(ToDoInputField).at(0).props().value).toBe(value);
    expect(wrapper.find(DeleteIcon).at(0).props()['$visible']).toBeTruthy()
    wrapper.find(DeleteIcon).at(0).simulate('click');
    expect(wrapper.find(ToDoInputField).at(0).props().value).toBe('');
  })

  it('handles input onChange & inserts toDo onKeyUp', () => {
    const wrapper = shallow(<ToDoInput onToDoInsert={onToDoInsert} />)
    wrapper.find(ToDoInputField).at(0).simulate('change', { target: { value } })
    expect(wrapper.find(ToDoInputField).at(0).props().value).toBe(value);

    wrapper.find(ToDoInputField).at(0).props().onKeyUp({ code: 'test' });
    expect(onToDoInsert).not.toHaveBeenCalled();
    expect(wrapper.find(ToDoInputField).at(0).props().value).toBe(value);

    wrapper.find(ToDoInputField).at(0).props().onKeyUp({ code: ENTER_KEY });
    expect(onToDoInsert).toHaveBeenCalled();
    expect(onToDoInsert).toHaveBeenCalledWith(value);
    expect(wrapper.find(ToDoInputField).at(0).props().value).toBe('');
  })

  it('shows ExpandIcon', () => {
    const wrapper = shallow(<ToDoInput />)
    expect(wrapper.find(ExpandIcon).length).toBe(1);
    expect(wrapper.find(ExpandIcon).props()['$visible']).toBeFalsy();

    expect(wrapper.find(HideIcon).length).toBe(0);
  })

  it('shows HideIcon', () => {
    const wrapper = shallow(<ToDoInput isExpanding={isExpanding} toggleListView={toggleListView} hasToDos={true} />)
    expect(wrapper.find(ExpandIcon).length).toBe(0);
  
    expect(wrapper.find(HideIcon).length).toBe(1);
    expect(wrapper.find(HideIcon).props()['$visible']).toBeTruthy();
    wrapper.find(HideIcon).at(0).simulate('click')
    expect(toggleListView).toHaveBeenCalled();
  })
})