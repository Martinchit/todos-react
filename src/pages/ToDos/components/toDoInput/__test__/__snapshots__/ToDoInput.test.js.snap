// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoInput renders correctly 1`] = `ShallowWrapper {}`;
